src/handler/active_requests.rs

```diff
@@ -1,69 +1,133 @@
+        {
+            Some(index) => index,
+            None => {
+                // if nonce req is missing, reinsert remaining requests into mapping
+                if !requests.is_empty() {
+                    self.active_requests_mapping
+                        .insert(node_address.clone(), requests);
                 }
-            },
-            None => None,
+                return Err(ActiveRequestsError::InvalidState);
+            }
+        };
+        let req = requests.remove(index);
+        if !requests.is_empty() {
+            self.active_requests_mapping
+                .insert(node_address.clone(), requests);
         }
+        Ok((node_address, req))
     }
 
-    pub fn remove(&mut self, node_address: &NodeAddress) -> Option<RequestCall> {
-        match self.active_requests_mapping.remove(node_address) {
-            Some(request_call) => {
-                // Remove the associated nonce mapping.
-                match self
-                    .active_requests_nonce_mapping
-                    .remove(request_call.packet().message_nonce())
-                {
-                    Some(_) => Some(request_call),
-                    None => {
-                        debug_unreachable!("A matching nonce mapping doesn't exist");
-                        error!("A matching nonce mapping doesn't exist");
-                        None
-                    }
-                }
-            }
-            None => None,
+    // Remove all requests associated with a node.
+    pub fn remove_requests(
+        &mut self,
+        node_address: &NodeAddress,
+    ) -> Result<Vec<RequestCall>, ActiveRequestsError> {
+        let requests = self
+            .active_requests_mapping
+            .remove(&node_address)
+            .ok_or_else(|| ActiveRequestsError::InvalidState)?;
+        for req in &requests {
+            self.active_requests_nonce_mapping
+                .remove(req.packet().message_nonce());
         }
+        Ok(requests)
+    }
+
+    // Remove a single request identified by its id.
+    pub fn remove_request(
+        &mut self,
+        node_address: &NodeAddress,
+        id: &RequestId,
+    ) -> Result<RequestCall, ActiveRequestsError> {
+        let reqs = self
+            .active_requests_mapping
+            .get(node_address)
+            .ok_or_else(|| ActiveRequestsError::InvalidState)?;
+        let index = reqs
+            .iter()
+            .position(|req| {
+                let req_id: RequestId = req.id().into();
+                &req_id == id
+            })
+            .ok_or_else(|| ActiveRequestsError::InvalidState)?;
+        let nonce = reqs
+            .get(index)
+            .ok_or_else(|| ActiveRequestsError::InvalidState)?
+            .packet()
+            .message_nonce()
+            .clone();
+        // Remove the associated nonce mapping.
+        let (_, request_call) = self.remove_by_nonce(&nonce)?;
```

Comment:
it would be ideal to do this here directly since this involves further checks and accesses already done here